### DISCUSSION TITLE
@OnFlush decorator

### DISCUSSION THREAD

**Discussion by xenoterracide** [mikro-orm/mikro-orm#3991] [Ideas]
I have

```ts
  onFlush(args: FlushEventArgs): void | Promise<void> {
    args.uow.getPersistStack().forEach((entity) => {
      if (entity instanceof AggregateRoot) {
        entity.commit();
      }
    });
  }
```

but it feels like this could easily be

```ts
@OnFlush()
onFlush(): void {
   this.commit(); // currently provided by a 3rd party parent class AggregateRoot
}
```

**Comment by B4nan** [comment]
Flush events are fired once per flush call, they are not bound to entity type anyhow. You don't flush entity, you flush entity manager. Technically, flush is **not** an entity lifecycle event, entities can get created, updated, or removed - not flushed.

**Reply by xenoterracide** [reply]
yes, but I simply want an action taken when flush happens, I guess I'm not understanding why this is a bad idea? it would certainly make my code less complicated (commit as the name of a method is stupid, I kind of hate the nestjs cqrs module).

note: not exactly sure my current code is 100% what it needs to be either, haven't determined if it should happen on something other than persist.

**Reply by B4nan** [reply]
You want a method to be triggered on an entity instance, but flush event is not bound to some entity instance, it is bound to the entity manager. All the hooks are about running code when something happens internally, e.g. when we compute a changeset for an entity - but flush does not happen internally for every entity, there is no place to add such hook. 

Sure we could iterate over all the changesets or something, but it simply feels a bit wrong to me. Note that if you want to do changes in the `onFlush` hook, you might need to recompute the changesets manually as described in the docs - the changesets are already computed.

Idk what's your use case or what is inside that `commit` method, but iterating over persist stack feels wrong to me as well, you will have the whole identity map there, every managed entity is automatically added to persist stack and checked for changes on each flush.

**Reply by xenoterracide** [reply]
it fires "domain events", that are stored in an array onto an event bus. That might be the wrong way, then suggestions? in this case it doesn't matter too much other than potentially being inefficient.
